target/linux/ipq806x/base-files/lib/upgrade/openmesh.sh

```diff
@@ -0,0 +1,87 @@
+# The OpenMesh devices come with 2 main partitions - while one is active
+# sysupgrade will flash the other. The boot order is changed to boot the
+# newly flashed partition. If the new partition can't be booted due to
+# upgrade failures the previously used partition is loaded.
+
+platform_do_upgrade_openmesh() {
+	local tar_file="$1"
+	local restore_backup
+	local primary_kernel_mtd
+
+	local setenv_script="/tmp/fw_env_upgrade"
+
+	local kernel_mtd="$(find_mtd_index $PART_NAME)"
+	local kernel_offset="$(cat /sys/class/mtd/mtd${kernel_mtd}/offset)"
+	local total_size="$(cat /sys/class/mtd/mtd${kernel_mtd}/size)"
+
+	# switch to second kernel when flashing to secondary flash region
+	local next_boot_part="1"
+	case "$(board_name)" in
+	openmesh,a42)
+		primary_kernel_mtd=8
+		;;
+	*)
+		echo "failed to detect primary kernel mtd partition for board"
+		return 1
+		;;
+	esac
+	[ "$kernel_mtd" = "$primary_kernel_mtd" ] || next_boot_part="2"
+
+	local board_dir=$(tar tf $tar_file | grep -m 1 '^sysupgrade-.*/$')
+	board_dir=${board_dir%/}
+
+	local kernel_length=$(tar xf $tar_file ${board_dir}/kernel -O | wc -c)
+	local rootfs_length=$(tar xf $tar_file ${board_dir}/root -O | wc -c)
+	# rootfs without EOF marker
+	rootfs_length=$((rootfs_length-4))
+
+	local kernel_md5=$(tar xf $tar_file ${board_dir}/kernel -O | md5sum); kernel_md5="${kernel_md5%% *}"
+	# md5 checksum of rootfs with EOF marker
+	local rootfs_md5=$(tar xf $tar_file ${board_dir}/root -O | dd bs=1 count=$rootfs_length | md5sum); rootfs_md5="${rootfs_md5%% *}"
+
+	#
+	# add tar support to get_image() to use default_do_upgrade() instead?
+	#
+
+	# take care of restoring a saved config
+	[ "$SAVE_CONFIG" -eq 1 ] && restore_backup="${MTD_CONFIG_ARGS} -j ${CONF_TAR}"
+
+	# write concatinated kernel + rootfs to flash
+	tar xf $tar_file ${board_dir}/kernel ${board_dir}/root -O | \
+		mtd $restore_backup write - $PART_NAME
+
+	# prepare new u-boot env
+	if [ "$next_boot_part" = "1" ]; then
+		echo "bootseq 1,2" > $setenv_script
+	else
+		echo "bootseq 2,1" > $setenv_script
+	fi
+
+	printf "kernel_size_%i 0x%08x\n" $next_boot_part $kernel_length >> $setenv_script
+	printf "vmlinux_start_addr 0x%08x\n" ${kernel_offset} >> $setenv_script
+	printf "vmlinux_size 0x%08x\n" ${kernel_length} >> $setenv_script
+	printf "vmlinux_checksum %s\n" ${kernel_md5} >> $setenv_script
+
+	printf "rootfs_size_%i 0x%08x\n" $next_boot_part $((total_size-kernel_length)) >> $setenv_script
+	printf "rootfs_start_addr 0x%08x\n" $((kernel_offset+kernel_length)) >> $setenv_script
+	printf "rootfs_size 0x%08x\n" ${rootfs_length} >> $setenv_script
+	printf "rootfs_checksum %s\n" ${rootfs_md5} >> $setenv_script
+
+	# store u-boot env changes
+	fw_setenv -s $setenv_script || {
+		echo "failed to update U-Boot environment"
+		return 1
+	}
+}
+
+# make sure we got uboot-envtools and fw_env.config copied over to the ramfs
+# create /var/lock for the lock "fw_setenv.lock" of fw_setenv
+platform_add_ramfs_ubootenv()
```

Comment:
This shouldn't be required. `target/linux/ipq806x/base-files/lib/upgrade/platform.sh` has already the required commands to copy `fw_*` to the ramfs:

```
RAMFS_COPY_BIN='fw_printenv fw_setenv'
RAMFS_COPY_DATA='/etc/fw_env.config /var/lock/fw_printenv.lock'
```

Reply:
Looks like it should work but it didn't work for me (stopped because /var/lock/fw_printenv.lock was not available). Can check later what is wrong

Reply:
Yes please check why it doesn't work. If it is a bug, other images within the target are affected as well.

Reply:
This is what I see during an upgrade without the copying

    root@OpenWrt:/# sysupgrade /tmp/openwrt-ipq806x-openmesh_a42-squashfs-sysupgrade.bin
    Saving config files...
    Commencing upgrade. All shell sessions will be closed now.
    Watchdog handover: fd=3
    - watchdog -
    killall: telnetd: no process killed
    Sending TERM to remaining processes ... ntpd ubusd logd netifd odhcpd dnsmasq
    Sending KILL to remaining processes ...
    Switching to ramdisk...
    Performing system upgrade...
    2883584+0 records in
    2883584+0 records out
    Unlocking inactive ...

    Writing from <stdin> to inactive ...
    Appending jffs2 data from /tmp/sysupgrade.tgz to inactive..                                                                                                                                                                                                                     
    Error opening lock file /var/lock/fw_printenv.lock
    failed to update U-Boot environment
    Upgrade completed
    Rebooting system...

The system will then reboot and return to the old flash region (the one which wasn't updated). It looks to me like the binary was copied but for some reason the /var/lock/fw_printenv.lock is not working. And I also found that linksys.sh is manually creating this file:

	mkdir -p /var/lock
	local part_label="$(linksys_get_target_firmware)"
	touch /var/lock/fw_printenv.lock

So I am not the only one experiencing this kind of problem. My assumption is now that:

1. fw_printenv.lock doesn't exist yet in /var/lock
2. sysupgrade is started
3. install_file doesn't find the source file
4. install_file skips the creation of /var/log

Just check the code of install_file

    install_file() { # <file> [ <file> ... ]
    	local target dest dir
    	for file in "$@"; do
    		if [ -L "$file" ]; then
    			target="$(readlink -f "$file")"
    			dest="$RAM_ROOT/$file"
    			[ ! -f "$dest" ] && {
    				dir="$(dirname "$dest")"
    				mkdir -p "$dir"
    				ln -s "$target" "$dest"
    			}
    			file="$target"
    		fi
    		dest="$RAM_ROOT/$file"
    		[ -f "$file" -a ! -f "$dest" ] && {
    			dir="$(dirname "$dest")"
    			mkdir -p "$dir"
    			cp "$file" "$dest"
    		}
    	done
    }

The last check `[ -f "$file" -a ! -f "$dest" ]` will not be started when the src file doesn't exist. And this is the reason why the `mkdir -p` in this block is not started.

I've tested this now by changing the function to only create the directory:

    # create /var/lock for the lock "fw_setenv.lock" of fw_setenv
    # the rest is copied using ipq806x's RAMFS_COPY_BIN and RAMFS_COPY_DATA
    platform_add_ramfs_ubootenv()
    {
            mkdir -p $RAM_ROOT/var/lock
    }
    append sysupgrade_pre_upgrade platform_add_ramfs_ubootenv
bootenv

This version now work perfectly fine (and I've added this version for now to my PR).

@NeoRaider you've worked on that code a while back. Maybe you have an opinion how this should work correctly.

Right now, the users of fw_setenv (allnet openmesh and linksys) seem to have a private `mkdir -p $RAMFS/var/lock` in their scripts. Either using a variant of `platform_add_ramfs_ubootenv` or a simple `mkdir -p` somewhere before they call fw_setenv

Reply:
We could simply add the `mkdir -p $RAM_ROOT/var/lock` to switch_to_ramfs() in `/lib/upgrade/stage2` rather than duplicating it in every target that needs it.

Reply:
I would love if we could somehow share the creation of these essential volatile directories with the mkdir commands in `/etc/init.d/boot` - maybe fold the mkdir commands there into a `/lib/function.sh` procedure we can call both from init and sysupgrade stage2?